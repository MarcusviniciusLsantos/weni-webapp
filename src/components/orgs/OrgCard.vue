<template>
  <div class="org-card" @click="$emit('enter')">
    <div>
      <h2 class="name">{{ name }}</h2>

      <p class="description">
        {{ description }}
      </p>

      <div v-if="plan" :class="['tag', plan]">
        {{ $t(`billing.payment.plans.${plan}.title`) }}
      </div>
    </div>

    <div v-if="$slots.options">
      <unnnic-dropdown
        @click.prevent
        :open.sync="isOptionsOpen"
        class="unnnic-dropdown"
      >
        <unnnic-icon
          slot="trigger"
          class="menu-icon"
          icon="navigation-menu-vertical-1"
          size="sm"
          :scheme="isOptionsOpen ? 'neutral-cloudy' : 'neutral-clean'"
        ></unnnic-icon>

        <slot name="options"></slot>
      </unnnic-dropdown>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    name: String,
    description: String,
    plan: String,
  },

  data() {
    return {
      isOptionsOpen: false,
    };
  },
};
</script>

<style lang="scss" scoped>
@import '~@weni/unnnic-system/src/assets/scss/unnnic.scss';

.org-card {
  cursor: pointer;
  display: flex;
  column-gap: $unnnic-spacing-sm;
  justify-content: space-between;

  outline-style: solid;
  outline-color: $unnnic-color-neutral-cleanest;
  outline-width: $unnnic-border-width-thinner;
  outline-offset: -$unnnic-border-width-thinner;

  background-color: $unnnic-color-background-white;
  padding: $unnnic-spacing-md;
  border-radius: $unnnic-border-radius-md;

  &:hover {
    box-shadow: $unnnic-shadow-level-near;
  }

  .name {
    color: $unnnic-color-neutral-black;

    font-family: $unnnic-font-family-secondary;
    font-weight: $unnnic-font-weight-bold;
    font-size: $unnnic-font-size-title-md;
    line-height: $unnnic-font-size-title-md + $unnnic-line-height-md;

    margin: 0;
    margin-bottom: $unnnic-spacing-ant;
  }

  .description {
    color: $unnnic-color-neutral-dark;

    font-family: $unnnic-font-family-secondary;
    font-weight: $unnnic-font-weight-regular;
    font-size: $unnnic-font-size-body-lg;
    line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;

    margin: 0;
  }

  .tag {
    display: inline-block;
    user-select: none;

    font-family: $unnnic-font-family-secondary;
    font-weight: $unnnic-font-weight-regular;
    font-size: $unnnic-font-size-body-md;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;

    margin-top: $unnnic-spacing-sm;
    padding: $unnnic-spacing-nano $unnnic-spacing-ant;
    border-radius: $unnnic-border-radius-pill;

    $plan-colors: 'trial' $unnnic-color-aux-blue-500,
      'scale' $unnnic-color-aux-orange-500,
      'advanced' $unnnic-color-aux-purple-500,
      'enterprise' $unnnic-color-aux-green-500;

    @each $name, $color in $plan-colors {
      &.#{$name} {
        color: $color;
        background-color: rgba($color, $unnnic-opacity-level-extra-light);
      }
    }
  }

  .menu-icon {
    display: block;
    margin-block: $unnnic-spacing-xs;
    user-select: none;
  }

  .unnnic-dropdown {
    :deep(.unnnic-dropdown__trigger .unnnic-dropdown__content) {
      margin-top: 0;
      padding: 0;
    }
  }
}
</style>
