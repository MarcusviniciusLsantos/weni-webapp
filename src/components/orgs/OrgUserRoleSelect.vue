<template>
  <UnnnicDropdown>
    <template
      v-if="type === 'button'"
      #trigger
    >
      <UnnnicButton
        class="weni-org-role__action__button"
        size="small"
        type="tertiary"
        iconRight="arrow-button-down-1"
      >
        {{ labelFor(modelValue) }}
      </UnnnicButton>
    </template>
    <template
      v-else-if="type === 'input'"
      #trigger
    >
      <UnnnicInput
        :label="$t('orgs.roles.permission')"
        size="md"
        iconRight="arrow-button-down-1"
        readonly
        :modelValue="labelFor(modelValue)"
        :disabled="disabled"
      />
    </template>

    <UnnnicDropdownItem
      v-for="roleOption in roleOptions"
      :key="roleOption.value"
      @click="onSelectRole(roleOption.value)"
    >
      <h4>{{ labelFor(roleOption.value) }}</h4>
      <span>
        {{ descriptionFor(roleOption.value) }}
      </span>
    </UnnnicDropdownItem>
  </UnnnicDropdown>
</template>

<script>
export default {
  name: 'OrgUserRoleSelect',

  props: {
    type: {
      required: true,
      type: String,
      validator(value) {
        return ['button', 'input'].includes(value);
      },
    },

    modelValue: {
      type: [String, Number],
    },

    disabled: Boolean,
  },

  data() {
    return {
      roleOptions: [
        { title: 'admin', value: 3 },
        { title: 'financial', value: 4 },
        { title: 'contributor', value: 2 },
      ],
    };
  },

  methods: {
    titleByValue(value) {
      return this.roleOptions.find(({ value: v }) => v == value)?.title;
    },
    labelFor(value) {
      return this.$t(`orgs.roles.${this.titleByValue(value)}`);
    },
    descriptionFor(value) {
      return this.$t(`orgs.roles.${this.titleByValue(value)}_description`);
    },
    onSelectRole(value) {
      this.$emit('update:model-value', value);
    },
  },
};
</script>
