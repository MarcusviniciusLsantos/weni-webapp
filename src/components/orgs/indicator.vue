<template>
  <div class="weni-step-indicator">
    <div
      v-for="step in steps"
      :class="{
        'weni-step-indicator__step__wrapper': true,
        'weni-step-indicator--has-separator': step > 1,
      }"
      :key="step"
    >
      <div
        v-if="step !== 1"
        :class="{
          'weni-step-indicator__step__separator': true,
          'weni-step-indicator__step__separator--active': step <= current,
        }"
      />
      <div
        :class="{
          'weni-step-indicator__step': true,
          'weni-step-indicator__step--active': step <= current,
        }"
      >
        <span class="weni-step-indicator__label">
          {{ nameForStep(step) }}
        </span>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    steps: {
      type: Number,
      required: true,
    },
    names: {
      type: Array,
      default: () => [],
    },
    current: {
      type: Number,
      default: 0,
    },
  },
  methods: {
    nameForStep(step) {
      return this.names[step - 1];
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~@weni/unnnic-system/src/assets/scss/unnnic.scss';
.weni-step-indicator {
  display: flex;
  align-items: center;
  width: 100%;

  &--has-separator {
    flex: 1;
  }

  &__step {
    border-radius: 50%;
    height: $unnnic-avatar-size-nano;
    width: $unnnic-avatar-size-nano;
    border: $unnnic-border-width-thick solid $unnnic-color-neutral-lightest;
    position: relative;

    &__wrapper {
      display: flex;
      align-items: center;
    }

    &--active {
      background-color: #00ded2;
      border: 4px solid rgba(0, 222, 210, 0.24);
      background-clip: content-box;
    }

    &__separator {
      border: $unnnic-border-width-thin solid $unnnic-color-neutral-lightest;
      margin: 0 $unnnic-inline-xs;
      border-radius: 3px;
      flex: 1;

      &--active {
        border: 2px solid $unnnic-color-brand-weni;
      }
    }
  }
  &__label {
    position: absolute;
    white-space: nowrap;
    text-align: center;
    transform: translateX(-50%);
    left: 50%;
    top: 100%;
    margin: $unnnic-inline-xs 0 0 0;
    color: $unnnic-color-neutral-cleanest;
    font-size: $unnnic-font-size-body-gt;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-medium;
  }
}
</style>
