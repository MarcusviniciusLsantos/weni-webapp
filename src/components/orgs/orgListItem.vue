<template>
  <unnnic-card-company
    :title="name"
    :tag="$t(`billing.${plan}.title`)"
    :description="description"
    :join-label="$t('orgs.join')"
    :options="options"
    :members="displayMembers"
    :members-description="
      remainingMembers ? $tc('orgs.remaining_members', remainingMembers) : null
    "
    @join="onSelectOrg"
  />
</template>

<script>
import image from '../../../src/assets/default-avatar.svg';

const ORG_ROLE_CONTRIBUTOR = 2;
const ORG_ROLE_ADMIN = 3;
const ORG_ROLE_FINANCIAL = 4;

export default {
  name: 'OrgListItem',
  props: {
    name: {
      type: String,
      default: null,
    },
    description: {
      type: String,
      default: null,
    },
    members: {
      type: Array,
      default: () => [],
    },
    canEdit: {
      type: Boolean,
      default: false,
    },
    canEditBilling: {
      type: Boolean,
      default: false,
    },
    role: {
      type: Number,
    },
    org: {
      type: Object,
    },
  },
  data() {
    return {
      highlighted: false,
      confirmOrganizationName: '',
    };
  },
  computed: {
    plan() {
      return this.org.organization_billing?.plan;
    },

    displayMembers() {
      if (!this.members) return [];

      return this.members.map((member) => ({
        name: this.getName(member),
        photo: member.photo_user || image,
      }));
    },
    remainingMembers() {
      if (!this.members || this.members.length <= 4) return 0;
      return this.members.length - 4;
    },

    options() {
      const billingArray = [
        {
          icon: 'currency-dollar-circle-1',
          title: this.$t('orgs.billing'),
          click: this.onSelectBilling,
        },
      ];

      if (this.role === ORG_ROLE_CONTRIBUTOR) {
        return [
          {
            icon: 'view-1-1',
            title: this.$t('orgs.view_members'),
            click: this.onView,
          },
        ];
      } else if (this.role === ORG_ROLE_FINANCIAL) {
        return billingArray;
      } else if (this.role === ORG_ROLE_ADMIN) {
        return [
          {
            icon: 'single-neutral-actions-1',
            title: this.$t('orgs.manage_members'),
            click: this.onManage,
          },
        ]
          .concat(billingArray)
          .concat([
            {
              icon: 'cog-1',
              title: this.$t('orgs.config'),
              click: this.onEdit,
            },
            {
              icon: 'delete-1-1',
              title: this.$t('orgs.delete.title'),
              click: this.openDeleteModal,
              scheme: 'feedback-red',
            },
          ]);
      }

      return [];
    },
  },
  methods: {
    openDeleteModal() {
      this.$emit('open-delete-confirmation');
    },

    getName(user) {
      const name = [user.first_name, user.last_name]
        .filter((name) => name)
        .join(' ');
      return name ? name : user.username;
    },

    onSelectOrg() {
      if (this.role === ORG_ROLE_FINANCIAL) {
        this.onSelectBilling();
      } else {
        this.$emit('select');
      }
    },
    onSelectBilling() {
      this.$emit('billing');
    },
    onEdit() {
      this.$emit('edit');
    },
    onManage() {
      this.$emit('manage');
    },
    onView() {
      this.$emit('view');
    },
  },
};
</script>
