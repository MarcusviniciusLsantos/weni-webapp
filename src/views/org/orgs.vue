<template>
  <div :class="['weni-orgs', `status-${organizationsStatus}`]">
    <div>
      <div class="weni-orgs__left">
        <div class="page-header">
          <UnnnicAvatarIcon
            v-if="organizationsStatus === 'empty'"
            class="weni-orgs__left__icon"
            icon="building-2-1"
            size="xl"
            scheme="weni-600"
          />

          <div>
            <h1>
              {{ $t('orgs.orgs') }}
            </h1>

            <p>
              {{
                $t(
                  organizationsStatus === 'empty'
                    ? 'orgs.orgs_description_empty'
                    : 'orgs.orgs_description',
                )
              }}
            </p>
          </div>

          <template v-if="error">
            <p>{{ $t('orgs.error_on_loading_orgs') }}</p>

            <UnnnicButton
              type="secondary"
              iconLeft="button-refresh-arrow-1"
              :disabled="organizationsStatus === 'loading'"
              @click="tryAgain()"
            >
              {{ $t('try_again') }}
            </UnnnicButton>
          </template>

          <template v-else>
            <RouterLink
              :to="{
                name: 'create_org',
              }"
            >
              <UnnnicButton
                class="create-org-button"
                type="primary"
                iconLeft="add-1"
              >
                {{ $t('orgs.add_org') }}
              </UnnnicButton>
            </RouterLink>
          </template>
        </div>
      </div>

      <template v-if="organizationsStatus !== 'empty'">
        <hr />

        <div class="unnnic-grid-span-5 weni-orgs__right">
          <div class="weni-orgs__list">
            <div class="filters">
              <UnnnicInput
                v-model="organizationName"
                iconLeft="search-1"
                size="md"
                :placeholder="$t('search')"
              ></UnnnicInput>

              <ListOrdinator
                v-model="$store.state.Org.orgs.ordering"
                :ordinators="['alphabetical', 'newer', 'older']"
              />
            </div>

            <OrgList
              ref="orgList"
              class="list-container"
              :filterName="organizationName"
            />
          </div>
        </div>
      </template>
    </div>
  </div>
</template>

<script>
import OrgList from '../../components/orgs/orgList.vue';
import ListOrdinator from '@/components/ListOrdinator.vue';
import { mapActions } from 'vuex';

export default {
  // eslint-disable-next-line vue/multi-word-component-names
  name: 'Orgs',
  components: {
    OrgList,
    ListOrdinator,
  },

  data() {
    return {
      error: false,
      organizationName: '',
    };
  },
  computed: {
    organizationsStatus() {
      if (
        this.$store.state.Org.orgs.status === 'complete' &&
        this.$store.state.Org.orgs.data.length === 0
      ) {
        return 'empty';
      }

      return this.$store.state.Org.orgs.status;
    },
  },

  watch: {
    organizationsStatus(status) {
      if (status === 'error') {
        this.error = true;
      } else if (status === 'loaded' || status === 'empty') {
        this.error = false;
      }
    },
  },

  mounted() {
    this.clearCurrentOrg();
    this.clearCurrentProject();
  },

  methods: {
    ...mapActions(['clearCurrentOrg', 'clearCurrentProject']),

    tryAgain() {
      this.$refs.orgList.reloadOrganizations();
    },
  },
};
</script>

<style lang="scss" scoped>
hr {
  width: 100%;
  margin-block: $unnnic-spacing-xl;
  border-width: 0;
  margin-block-start: $unnnic-spacing-xl - $unnnic-border-width-thinner;
  border-top: $unnnic-border-width-thinner solid $unnnic-color-neutral-cleanest;
}

.weni-orgs {
  display: flex;
  margin: 0 12.88%;
  flex: 1;
  display: flex;
  flex-direction: column;

  &.page {
    overflow: initial !important;
  }

  &.status-empty {
    justify-content: center;

    .page-header {
      display: block;

      p {
        margin: 0 0 $unnnic-spacing-stack-md 0;
      }

      .create-org-button {
        width: 15.625 * $unnnic-font-size;
      }
    }

    > div {
      .weni-orgs__left {
        margin: 0px auto;
        max-width: 25.8125rem;
        text-align: center;
        align-items: center;
      }
    }
  }

  a {
    text-decoration: none;
  }

  .unnnic-grid-lg {
    flex: 1;
    padding: 0;
    box-sizing: border-box;
    padding: 0 12.88%;
    padding-bottom: $unnnic-spacing-stack-xl - ($unnnic-border-width-thick * 2);
    border-bottom: $unnnic-border-width-thick * 2 solid $unnnic-color-brand-weni;
  }

  &__right {
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding-bottom: $unnnic-spacing-stack-lg;

    &::-webkit-scrollbar {
      display: none;
    }

    .filters {
      margin-bottom: $unnnic-spacing-stack-md;
      display: flex;
      flex-wrap: wrap;
      align-items: center;
      gap: $unnnic-spacing-stack-sm $unnnic-spacing-inline-md;

      .unnnic-form {
        flex: 1;
        min-width: 14rem;
      }
    }
  }

  &__list {
    // overflow-y: scroll;
    -ms-overflow-style: none; /* IE and Edge */
    scrollbar-width: none;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    // max-height: 100%;
    // height: 100%;

    // > * {
    //   margin-bottom: $unnnic-spacing-stack-xs;
    // }

    flex: 1;
    // overflow: overlay;
    // min-height: 4rem;

    $scroll-size: $unnnic-inline-nano;
    // padding-right: calc(#{$unnnic-inline-xs} + #{$scroll-size});
    width: 100%;

    &::-webkit-scrollbar {
      width: $scroll-size;
    }

    &::-webkit-scrollbar-thumb {
      background: $unnnic-color-neutral-clean;
      border-radius: $unnnic-border-radius-pill;
    }

    &::-webkit-scrollbar-track {
      background: $unnnic-color-neutral-soft;
      border-radius: $unnnic-border-radius-pill;
    }
  }

  &__left {
    font-family: $unnnic-font-family-primary;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: flex-start;
    margin: $unnnic-spacing-stack-lg 0;

    &__icon {
      margin: 0 0 $unnnic-spacing-stack-md 0;
    }

    .page-header {
      display: flex;
      align-items: center;
      justify-content: space-between;
      width: 100%;
      column-gap: 12%;

      h1 {
        font-family: $unnnic-font-family-primary;
        color: $unnnic-color-neutral-darkest;
        font-weight: $unnnic-font-weight-bold;
        font-size: $unnnic-font-size-title-lg;
        line-height: $unnnic-font-size-title-lg + $unnnic-line-height-md;

        margin: 0 0 $unnnic-spacing-sm 0;
      }

      p {
        font-family: $unnnic-font-family-secondary;
        color: $unnnic-color-neutral-dark;
        font-weight: $unnnic-font-weight-regular;
        font-size: $unnnic-font-size-body-lg;
        line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;

        margin: 0;
      }

      .create-org-button {
        width: 19.875 * $unnnic-font-size;
      }
    }
  }
}
</style>
