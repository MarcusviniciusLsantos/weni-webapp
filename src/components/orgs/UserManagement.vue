<template>
  <div>
    <div class="group">
      <template v-if="type === 'manage'">
        <UnnnicInputNext
          v-model="userSearch"
          :label="$t('orgs.roles.add_member')"
          :placeholder="$t('orgs.roles.add_member_placeholder')"
          :error="emailError"
          :disabled="loadingAddingUser || loading"
        />

        <div class="multiSelect">
          <OrgUserRoleSelect
            type="input"
            v-model="role"
            :disabled="loadingAddingUser || loading"
            :class="{ 'org__button-fix-margin': emailError }"
          />
        </div>

        <UnnnicButton
          @click="onSubmit"
          :disabled="
            loadingAddingUser ||
            loading ||
            !userSearch.trim().length ||
            !!emailError
          "
          :class="{ 'org__button-fix-margin': emailError }"
          type="primary"
          size="large"
          :style="{ flex: 1, minWidth: `8rem` }"
        >
          {{ $t('add') }}
        </UnnnicButton>
      </template>

      <template v-else-if="type === 'read'">
        <SearchUser
          :modelValue="searchName"
          @update:model-value="$emit('update:search-name', $event)"
          @reset="$emit('reset')"
          class="weni-org-permissions__input"
          :label="$t('orgs.create.user_search')"
          :placeholder="$t('orgs.create.user_search_description')"
        />
      </template>
    </div>

    <div class="users">
      <OrgRole
        v-for="(user, index) in users"
        :disabled="isMe(user) || user.disabledRole || type === 'read'"
        :role="user.role"
        :key="index"
        :email="user.email"
        :username="user.username"
        :name="isMe(user) ? $t('orgs.you') : user.name"
        :imageUrl="user.photo"
        :deleteTooltip="
          isMe(user) ? $t('orgs.users.leave') : $t('orgs.users.remove')
        "
        :canDelete="
          type === 'read' ? false : cannotDeleteMyUser ? !isMe(user) : true
        "
        :status="capitalize(user.status && $t(`status.${user.status}`))"
        @onChangeRole="onEdit($event, user)"
        @onDelete="onRemove(user)"
        class="user"
      />
      <InfiniteLoading
        v-if="!doNotFetch"
        @infinite="$emit('fetch-permissions', $event)"
      />
    </div>
  </div>
</template>

<script>
import { mapActions } from 'vuex';
import OrgRole from './orgRole.vue';
import InfiniteLoading from '../InfiniteLoading.vue';
import Unnnic from '@weni/unnnic-system';
import _ from 'lodash';
import orgs from '../../api/orgs';
import SearchUser from './searchUser.vue';
import OrgUserRoleSelect from './OrgUserRoleSelect.vue';

export default {
  components: {
    OrgRole,
    InfiniteLoading,
    SearchUser,
    OrgUserRoleSelect,
  },

  emits: ['update:users'],

  props: {
    type: String,

    users: {
      type: Array,
    },

    doNotFetch: {
      type: Boolean,
    },

    cannotDeleteMyUser: {
      type: Boolean,
    },

    loading: {
      type: Boolean,
    },

    offline: {
      type: Boolean,
    },

    org: {
      type: Object,
      default() {
        return {};
      },
    },

    alreadyAddedText: {
      type: String,
    },

    searchName: String,
  },

  data() {
    return {
      role: '3',

      userSearch: '',
      forceTooltipPressEnterOpen: true,

      loadingAddingUser: false,

      removingUser: null,
    };
  },

  mounted() {},

  computed: {
    emailError() {
      if (
        this.userSearch.trim().length &&
        !this.rules.email.test(this.userSearch.trim())
      ) {
        return this.$t('errors.invalid_email');
      }

      if (
        this.users.some(
          ({ email }) =>
            email.toLowerCase() === this.userSearch.trim().toLowerCase(),
        )
      ) {
        return this.$t('orgs.users.already_in');
      }

      return false;
    },
  },

  methods: {
    ...mapActions([
      'searchUsers',
      'leaveOrg',
      'removeAuthorization',
      'openModal',
    ]),

    capitalize: _.capitalize,

    isMe(user) {
      return user.username === this.$store.state.Account.profile.username;
    },

    onEdit(role, user) {
      if (this.offline) {
        this.$emit(
          'update:users',
          this.users.map((item) =>
            item.email === user.email ? { ...user, role } : item,
          ),
        );
      } else {
        this.$emit('change-role', {
          id: user.id,
          role,
        });
      }
    },

    clearUserFromChanges(user) {
      this.$emit(
        'update:users',
        this.users.filter((item) => item.username !== user.username),
      );
    },

    onRemove(user) {
      if (user.offline) {
        this.clearUserFromChanges(user);
      } else {
        this.removingUser = user.username;

        let title = '';
        let description = '';
        let validate = null;

        if (this.isMe(user)) {
          title = this.$t('orgs.leave.title');
          description = this.$t('orgs.leave_description');
          validate = {
            label: this.$t('orgs.leave.confirm_with_name', {
              name: this.org.name,
            }),
            placeholder: this.$t('orgs.leave.confirm_with_name_placeholder'),
            text: this.org.name,
          };
        } else {
          title = this.$t('orgs.remove_member');
          description = this.$t('orgs.remove_member_description', {
            user: user.name,
            org: this.org.name,
          });
        }

        this.openModal({
          type: 'confirm',
          data: {
            persistent: true,
            icon: 'alert-circle-1',
            scheme: 'feedback-red',
            title,
            description,
            validate,
            cancelText: this.$t('cancel'),
            confirmText: title,
            onConfirm: async (justClose, { setLoading }) => {
              setLoading(true);
              await this.removeRole();
              setLoading(false);

              justClose();
            },
          },
        });
      }
    },

    async removeRole() {
      const user = this.users.find(
        (user) => user.username === this.removingUser,
      );

      if (this.isMe(user)) {
        await this.onLeaveOrg(user.id);
        this.$emit('remove-user', this.removingUser);
        this.removingUser = null;
        return;
      }

      try {
        if (user.status === 'pending') {
          await orgs.deleteRequestPermission({
            organization: this.org.uuid,
            id: user.id,
          });
        } else {
          await this.removeAuthorization({
            orgId: this.org.uuid,
            username: user.id,
          });
        }

        this.clearUserFromChanges(user);

        this.openModal({
          type: 'alert',
          data: {
            icon: 'check_circle',
            scheme: 'feedback-green',
            title: this.$t('orgs.removed_member'),
            description: this.$t('orgs.removed_member_success', {
              user: this.removingUser,
            }),
          },
        });
      } catch (e) {
        Unnnic.unnnicCallModal({
          props: {
            text: this.$t('orgs.error'),
            description: this.$t('orgs.save_error'),
            scheme: 'feedback-red',
            icon: 'check_circle',
          },
        });
      } finally {
        this.$emit('remove-user', this.removingUser);
        this.removingUser = null;
      }
    },

    async onLeaveOrg(id) {
      try {
        await this.leaveOrg({
          orgId: this.org.uuid,
          id,
        });

        this.openModal({
          type: 'alert',
          data: {
            icon: 'check_circle',
            scheme: 'feedback-green',
            title: this.$t('orgs.users.left', { name: this.org.name }),
            description: this.$t('orgs.users.left_description'),
          },
        });

        this.$emit('finish');
      } catch (e) {
        Unnnic.unnnicCallModal({
          props: {
            text: this.$t('orgs.error'),
            description: this.$t('orgs.save_error'),
            scheme: 'feedback-red',
            icon: 'check_circle',
          },
        });
      }
    },

    async onSubmit() {
      this.loadingAddingUser = true;

      const email = this.userSearch.trim().toLowerCase();

      try {
        const { data } = await this.searchUsers({
          search: email,
        });

        const users = data.filter((user) => user.email === email);

        let addedUser = null;

        const role = this.role;

        if (users.length) {
          const [user] = users;

          addedUser = {
            id: user.id,
            uuid: Math.random(),
            name: [user.first_name, user.last_name]
              .filter((name) => name)
              .join(' '),
            email: user.email,
            photo: user.photo,
            role,
            username: user.username,
            offline: true,
          };
        } else {
          addedUser = {
            id: email,
            uuid: Math.random(),
            name: email,
            email: email,
            photo: null,
            role,
            username: email,
            offline: true,
            status: 'pending',
          };
        }

        if (this.offline) {
          this.$emit('update:users', this.users.concat(addedUser));
        } else {
          this.$emit('add', addedUser);
        }

        this.userSearch = '';
      } catch (error) {
        console.log(error);
      } finally {
        this.loadingAddingUser = false;
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.unnnic-dropdown {
  :deep(.unnnic-dropdown__trigger) {
    width: 100%;
  }

  :deep(.unnnic-dropdown__content) {
    min-width: calc(276px - 32px);

    a > span {
      width: 100%;
      display: block;
      font-weight: $unnnic-font-weight-light;
    }

    h4 {
      font-weight: $unnnic-font-weight-regular;
    }

    h4,
    a > span {
      margin: 0;
      line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
      font-size: $unnnic-font-size-body-md;
      color: $unnnic-color-neutral-dark;
    }
  }
}

.group {
  display: flex;
  margin-bottom: $unnnic-spacing-stack-md;
  align-items: flex-end;
  width: 100%;

  > div {
    width: 100%;
  }

  .unnnic-form {
    max-width: 280px;
  }

  .multiSelect {
    max-width: 196px;
  }

  .org__button-fix-margin {
    margin-bottom: $unnnic-spacing-stack-md;
  }

  > *:not(:last-child) {
    margin-right: $unnnic-spacing-stack-sm;
  }

  .flex-1 {
    flex: 1;
  }
}

.users {
  flex: 1;

  $scroll-size: $unnnic-inline-nano;
  padding-right: calc(#{$unnnic-inline-xs} + #{$scroll-size});
  width: 100%;

  &::-webkit-scrollbar {
    width: $scroll-size;
  }

  &::-webkit-scrollbar-thumb {
    background: $unnnic-color-neutral-clean;
    border-radius: $unnnic-border-radius-pill;
  }

  &::-webkit-scrollbar-track {
    background: $unnnic-color-neutral-soft;
    border-radius: $unnnic-border-radius-pill;
  }

  .user:not(:last-child) {
    margin-bottom: $unnnic-spacing-stack-sm;
  }
}

.normal-multiselect {
  :deep(.select-content) {
    min-width: 349px;
    z-index: 2;
    right: 0;
  }
}
</style>
