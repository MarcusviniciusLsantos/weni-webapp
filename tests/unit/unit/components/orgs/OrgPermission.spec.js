import { vi } from 'vitest';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';

import orgPermissions from '@/components/common/RightBar/orgPermissions.vue';
import UserManagement from '@/components/orgs/UserManagement.vue';
import i18n from '@/utils/plugins/i18n';
import { org, user } from '../../../__mocks__/';

const localVue = createLocalVue();
localVue.use(Vuex);

vi.mock('@/api/request.js', () => {});

import { unnnicCallAlert as mockUnnnicCallAlert } from '@weni/unnnic-system';

vi.mock('@weni/unnnic-system', () => ({
  unnnicCallAlert: vi.fn(),
  unnnicCallModal: vi.fn(),
}));

import orgs from '@/api/orgs';

vi.mock('@/api/orgs.js', () => ({
  default: {
    createRequestPermission: vi.fn(),
  },
}));

describe('orgPermissions.vue', () => {
  let wrapper;
  let state;
  let actions;
  let store;

  beforeEach(() => {
    state = {
      Org: {
        orgs: { data: [org] },
      },
    };

    actions = {
      getMembers: vi.fn(),
      changeAuthorization: vi.fn(),
      openModal: vi.fn(),
    };

    store = new Vuex.Store({
      state,
      actions,
    });

    wrapper = shallowMount(orgPermissions, {
      localVue,
      store,
      i18n,
      mocks: {
        $t: () => 'some specific text',
        Keycloak: vi.fn(),
      },
      stubs: {
        orgRole: true,
        SearchUser: true,
        InfiniteLoading: true,
        UserManagement,
      },
      propsData: {
        orgUuid: org.uuid,
      },
    });
  });

  it('renders a snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('test props', async () => {
    expect(wrapper.vm.org).toBe(org);
  });

  it('genericError()', () => {
    wrapper.vm.genericError();
    expect(actions.openModal).toHaveBeenCalled();
  });

  describe('changeRole()', () => {
    it('works as well', async () => {
      actions.changeAuthorization.mockImplementation(() => {
        return true;
      });

      await wrapper.vm.changeRole({ id: '123', role: 3 });

      expect(mockUnnnicCallAlert).toHaveBeenCalled();
    });

    it('got an error', async () => {
      actions.changeAuthorization.mockImplementation(() => {
        throw new Error('error fetching');
      });

      await wrapper.vm.changeRole({ id: '123', role: 3 });

      expect(actions.openModal).toHaveBeenCalled();
    });
  });

  describe('addMember()', () => {
    it('got an error', async () => {
      orgs.createRequestPermission.mockImplementation(() => {
        throw new Error('error fetching');
      });

      const spy = vi.spyOn(wrapper.vm, 'genericError');

      await wrapper.vm.addMember(user);

      expect(spy).toHaveBeenCalled();
    });

    // it('got an success', async () => {
    //   orgs.createRequestPermission.mockImplementation(() => {
    //     return { user, id: '321', status: 'pending' };
    //   });

    //   user.status = 'pending';

    //   await wrapper.vm.addMember(user);

    //   expect(user.id).toEqual('321');
    // });
  });

  describe('fetchPermissions', () => {
    it('got an error', async () => {
      actions.getMembers.mockImplementation(() => {
        throw new Error('error fetching');
      });

      const state = {
        error: vi.fn(),
        loaded: vi.fn(),
      };

      const spyError = vi.spyOn(state, 'error');
      const spyLoaded = vi.spyOn(state, 'loaded');

      await wrapper.vm.fetchPermissions(state);

      expect(spyError).toHaveBeenCalled();
      expect(spyLoaded).toHaveBeenCalled();
      expect(wrapper.vm.loading).toBeFalsy();
    });

    it('got finished with complete true', async () => {
      actions.getMembers.mockImplementation(() => {
        return { data: { results: [user] } };
      });

      const state = {
        error: vi.fn(),
        loaded: vi.fn(),
        complete: vi.fn(),
      };

      const spyComplete = vi.spyOn(state, 'complete');

      await wrapper.vm.fetchPermissions(state);

      expect(spyComplete).toHaveBeenCalled();
      expect(wrapper.vm.loading).toBeFalsy();
    });
  });
});
