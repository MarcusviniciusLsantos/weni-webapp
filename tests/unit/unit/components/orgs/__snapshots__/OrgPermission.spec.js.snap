// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`orgPermissions.vue > renders a snapshot 1`] = `
VueWrapper {
  "_emitted": {},
  "_emittedByOrder": [],
  "isFunctionalComponent": undefined,
}
`;

exports[`orgPermissions.vue renders a snapshot 1`] = `
<div class="weni-org-permissions">
  <div style="display: flex; flex-direction: column; flex: 1;">
    <div class="group">
      <unnnicinputnext-stub value="" label="some specific text" size="md" type="normal" nativetype="text" placeholder="some specific text"></unnnicinputnext-stub>
      <div class="multiSelect">
        <orguserroleselect-stub type="input" value="3" class=""></orguserroleselect-stub>
      </div>
      <unnnicbutton-stub size="large" type="primary" disabled="true" class="" style="flex: 1; min-width: 8rem;">
        some specific text
      </unnnicbutton-stub>
    </div>
    <div class="users">
      <infiniteloading-stub></infiniteloading-stub>
    </div>
  </div>
</div>
`;
